server/master/src/main/java/org/apache/accumulo/master/TimeoutTaskExecutor.java

```diff
@@ -0,0 +1,249 @@
+ * tasks that error.
+ *
+ * This class uses an underlying fixed thread pool to schedule the submitted tasks. Once a task is submitted, the desired end time for the task is recorded and
+ * used to determine the timeout for the task's associated {@link Future}.
+ *
+ * The timeout will not be exact as the start time is recorded prior to submitting the {@link Callable}. This may result in an effective timeout that is
+ * slightly smaller than expected. The timeout used during initialization should be adjusted accordingly.
+ *
+ * The {@link TimeoutTaskExecutor} itself is not a thread-safe class. Only a single thread should submit tasks and complete them.
+ *
+ * @param <T> The return type for the corresponding Callable.
+ * @param <C> The type of Callable submitted to this executor.
+ */
+@NotThreadSafe
+public class TimeoutTaskExecutor<T, C extends Callable<T>> implements AutoCloseable {
+
+  private final static Logger log = LoggerFactory.getLogger(TimeoutTaskExecutor.class);
+
+  private final long timeoutInNanos;
+  private final ExecutorService executorService;
+  private final BlockingQueue<WrappedTask> startedTasks;
+  private final List<WrappedTask> wrappedTasks;
+
+  private SuccessCallback<T,C> successCallback;
+  private ExceptionCallback<C> exceptionCallback;
+  private TimeoutCallback<C> timeoutCallback;
+
+  /**
+   * Constructs a new TimeoutTaskExecutor that will use the given number of worker threads and timeout. Takes an expected number of Callables to initialize the
+   * underlying data structures appropriately.
+   * <p>
+   * If the expectedNumCallables is sized too small, this executor will block on calls to submit() once the internal queue is full.
+   *
+   * @param numThreads           The number of threads to use.
+   * @param timeoutInMillis      The timeout for each task in milliseconds.
+   * @param expectedNumCallables The expected number of callables you will schedule. Note this is used for an underlying BlockingQueue. If sized too small this will cause blocking
+   *                             when calling submit().
+   * @throws IllegalArgumentException If numThreads is less than 1 or expectedNumCallables is negative.
+   */
+  public TimeoutTaskExecutor(int numThreads, long timeoutInMillis, int expectedNumCallables) {
+    Preconditions.checkArgument(numThreads >= 1, "Number of threads must be at least 1.");
+    Preconditions.checkArgument(expectedNumCallables >= 0, "The expected number of callables must be non-negative.");
+
+    this.executorService = Executors.newFixedThreadPool(numThreads);
+    this.startedTasks = new ArrayBlockingQueue<>(expectedNumCallables);
+    this.timeoutInNanos = TimeUnit.MILLISECONDS.toNanos(timeoutInMillis);
+    this.wrappedTasks = new ArrayList<>(expectedNumCallables);
+  }
+
+  /**
+   * Submits a new task to the executor.
+   *
+   * @param callable Task to run
+   */
+  public void submit(C callable) {
```

Comment:
It would be nice to add a sanity check that makes this fail if complete() was called or is being called()

Reply:
Agreed. Will fix.